[1/2] `WCProductVariationModel` to Room: working app

wzieba opened the pull request:
- #13972
- #13982 

<!-- Remember about a good descriptive title. -->
### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCProductVariationModel` from WellSql to Room. It's similar to https://github.com/woocommerce/woocommerce-android/pull/13935, but this time the migration has smaller scope, as `WCProductVariationModel` is not used in that many places.
- The real diff of this PR is `+343` / `-396`, excluding JSON schema file
- Most changes are related to the fact of changing `WCProductVariationModel` to a data class with immutable members


### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Please test using #13982 

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

<img width="1458" alt="Screenshot 2025-04-30 at 13 32 21" src="https://github.com/user-attachments/assets/7d10d6dc-d270-430f-8268-bbcf2c16b9e4" />

- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Migrate `WCProductVariationModel` to Room

`:libs:fluxc-plugin:assembleDebug` passes

wzieba committed:
Migrate `getVariations` and `observeVariations` to Room's DAO

wzieba committed:
Migrate `getVariation` to Room's DAO

wzieba committed:
Migrate `upsertVariation`(s) to Room's DAO

wzieba committed:
Migrate `deleteVariation` to Room's DAO

wzieba committed:
Remove unused observability wrappers for variations in `ProductSqlUtils`

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ProductVariationsDao</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Adjust `WooCommerce` module to WCProductVariationModel changes: update to app model mapper, make query methods suspendable

wzieba committed:
Fix operations on ProductVariationsDao by passing **local** site id

To reduce chance of similar mistake in the future, I propose using `LocalId/RemoteId` wrappers.

wzieba committed:
Fix `ProductTestUtils` from testFixtures

To fix lint task

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/details/ProductDetailCardBuilder.kt:
```diff
@@ -518,9 +519,11 @@ class ProductDetailCardBuilder(
-                                        variationRepository.getProductVariationList(product.remoteId).all {
-                                            (it as? SubscriptionProductVariation)?.subscriptionDetails
-                                                ?.supportsOneTimeShipping ?: false
+                                        runBlocking {
+                                            variationRepository.getProductVariationList(product.remoteId).all {
+                                                (it as? SubscriptionProductVariation)?.subscriptionDetails
+                                                    ?.supportsOneTimeShipping ?: false
+                                            }
```
Sure, but it's intentional like in the other PR (https://github.com/woocommerce/woocommerce-android/pull/13935#discussion_r2050349317) - the goal of this PR is not to fix issues related to threading.

wzieba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/ProductVariation.kt:
```diff
@@ -134,49 +134,71 @@ open class ProductVariation(
+            }
+        }
+
+        fun attributesToJson(): String {
+            val jsonArray = JsonArray()
+            attributes.forEach { variantOption ->
+                JsonObject().apply {
+                    addProperty("id", variantOption.id)
+                    addProperty("name", variantOption.name)
+                    addProperty("option", variantOption.option)
+                }.also { jsonArray.add(it) }
+            }
+            return jsonArray.toString()
+        }
```
this replaces `addVariant` method. I like it doesn't need `gson` object (we shouldn't use `gson` inside models, nor create instance of `gson` per model), but we also loose synchronization of fields with `ProductVariantOption`. I think it's acceptable, because I think it's unlikely this model will change.

https://github.com/woocommerce/woocommerce-android/pull/13972/files#diff-a2f8ea1eb59520371a927db507f69e8a3da9de51a52c6cb1665a1bb2b56cbe17L108-L115

ParaskP7 replied:
I spend much more time that I wanted to this in trying to make sure to test this change, making sure it doesn't (somehow) change in functionality. So much complexity for nothing (the before). Many thanks for the creating this new `attributesToJson()` function that replaces the old, much better. 🙇

FYI: I personally verified the before and after, by reviewing that the resulting `attributes` string is matching the expectation by debugging the app and database. ✅ 

> ...but we also loose synchronization of fields with ProductVariantOption. I think it's acceptable, because I think it's unlikely this model will change.

PS: I'll ask the product team for a second opinion on that. 👍

wzieba commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductVariationModel.kt:
```diff
@@ -22,77 +17,58 @@ typealias VariationAttributes = List<ProductVariantOption>
-    }
+@Entity(
+    tableName = "ProductVariationEntity",
+    primaryKeys = ["localSiteId", "remoteProductId", "remoteVariationId"],
```
This `primaryKeys` introduces a new behavior of `UNIQUE` constraint, which wasn't present before, yet I think it's expected.

wzieba committed:
Fix missing buildkite test collector support when JaCoCo fails

In case when `jacocoTestReport` task fails, our script was not moving test results to the correct directory, as it was exiting on error before collection.

wzieba committed:
Run `jacocoTestReport` on the same Gradle invocation. Upload code coverage only on success tests

This will reduce chances of `testJalapenoDebugUnitTest` task being not up to date when running `jacocoTestReport`, which caused running unit tests again, without any reason.

wzieba committed:
Add missing modules test tasks to JacocoReport dependencies

To address issues like: Reason: Task ':jacocoTestReport' uses this output of task ':libs:apifaker:testDebugUnitTest' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

wzieba committed:
Print "Uploading code coverage" in a new line

wzieba committed:
Open `Testing` by default

There are two reasons for this:
- with the current setup, we were opening "Copying test logs for test collector" if tests failed, because this was the last collapsed regular group (docs: https://buildkite.com/docs/pipelines/configure/managing-log-output#collapsing-output) and it was confusing: tests did fail, not copying test logs
- developers are much more interested in details of `Testing` step than (empty) details of "Copying test logs" so it makes sense to open former by default

wzieba committed:
echo report tests status group in a new line

wzieba committed:
Actually print some section headers in a new line

wzieba committed:
Drop `WCProductVariationModel` from WellSql

wzieba committed:
Make `ProductVariationsDao` internal

wzieba committed:
Merge branch 'trunk' into migrate_product_variation_to_room

wzieba committed:
Merge pull request #13986 from woocommerce/fix_tests_collection_on_jacoco_fail

Fix Buildkite test collector support when tests execution fail

wzieba committed:
Merge branch 'feature/product_variation_to_room' into migrate_product_variation_to_room

ParaskP7 commented:
FYI: I am going to merge this PR into its feature branch, then merge #13982 as well, and finally open a single PR with all the changes. I will then update with latest `trunk`, apply some of my own (review) change, test and request a final review/test from the product team.

Why? Contrarily to `WCProductModel`, which also had multiple PRs and a feature branch, I think this PR isn't as involved, plus there already exists a precedent of such a change, which make reviewing/testing this as a whole much more straightforward.

Cc @wzieba

ParaskP7 commented:
FYI: As I was testing this change and running unit tests locally (in CLI), using the `./gradlew testDebugUnitTest` was consistently failing on `1087` test with the below (example) stack-trace on `ByteBuddyAgent`, similar to the Java 11 versus 17 problem we had with Mockito a while back:

```
    UploadStoreUnitTest > testNumberOfAutoUploadsAttemptsCounter FAILED
        java.lang.IllegalStateException at UploadStoreUnitTest.java:46
            Caused by: java.lang.IllegalStateException at PluginInitializer.java:56
                Caused by: java.lang.reflect.InvocationTargetException at NativeConstructorAccessorImpl.java:-2
                    Caused by: org.mockito.exceptions.base.MockitoInitializationException at InlineDelegateByteBuddyMockMaker.java:260
                        Caused by: java.lang.IllegalStateException at ByteBuddyAgent.java:706
```

The interesting thing is that tests run successfully when I split `testDebugUnitTest` into multiple modules, running them as:
    - `./gradlew libs:fluxc:testDebugUnitTest`
    - `./gradlew libs:fluxc-plugin:testDebugUnitTest`
    - `./gradlew libs:fluxc-tests:testDebugUnitTest`

What's more interesting, after me doing the above, even `./gradlew testDebugUnitTest` start working as expected (I also used the `--rerun-tasks` flag to make sure this unit test task runs again).

Any idea on what might the problem be, did it happen on your side? Is it maybe related to `testFixtures`, I can't think of any other major change that could start causing this? 🤔

PS: Also, I had no trouble running those unit tests from within the IDE (AS).
